File: genesis/app/artifact.py
#!/usr/bin/env python3
from os import environ
from jinja2 import Environment, PackageLoader, select_autoescape, StrictUndefined
from logging import getLogger
from io import BytesIO
from tarfile import TarInfo, DIRTYPE, open as tar_open
from time import time
from subprocess import Popen, PIPE
from hashlib import sha256
from .config import Config
log = getLogger(__name__)

class Artifact(object):
    def __init__(self,args,modes):
        self.__devel = args.devel
        self.__modes = modes
        self.__cfg = Config().main_config()
        self.__env = Environment(
            loader = PackageLoader("genesis"),
            undefined = StrictUndefined,
            autoescape = select_autoescape(),
        )
        self.__tpl_critcl_yaml = self.__env.get_template("crictl.yaml")
        self.__tpl_opt_bin_e = self.__env.get_template("opt-bin-e.yaml")
        self.__tpl_flatcar_update_conf = self.__env.get_template("flatcar-update.conf")
        self.__tpl_k3s_config_yaml = self.__env.get_template("k3s-config.yaml")
        self.__tpl_k3s_override_conf = self.__env.get_template("k3s-override.conf")
        self.__tpl_systemd_i12e_conf = self.__env.get_template("systemd-i12e.conf")
        self.__tpl_artifact_tune = self.__env.get_template("artifact-tune.sh")
        self.__tpl_nftables_conf = self.__env.get_template("nftables.conf")
        self.__tpl_nftables_service = self.__env.get_template("nftables.service")
        self.__time = time()

    def __tarinfo(self,fname):
        tinfo = TarInfo(name=fname)
        tinfo.mode = 0o644
        tinfo.mtime = self.__time
        tinfo.uname = "root"
        tinfo.gname = "root"
        return tinfo

    def __folders(self,tar):
        folders = [
            ("etc/i12e",0o700,"root","root"),
            ("etc/i12e/flags",0o700,"root","root"),
            ("etc/i12e/k3s",0o700,"root","root"),
            ("etc/systemd/system/k3s.service.d",0o755,"root","root"),
            ("etc/systemd/system.conf.d",0o755,"root","root"),
            ("opt/libexec",0o755,"root","root"),
            ("opt/libexec/i12e",0o755,"root","root"),
        ]
        for f in folders:
            ti = self.__tarinfo(f[0])
            ti.mode = f[1]
            ti.type = DIRTYPE
            ti.uname = f[2]
            ti.gname = f[3]
            tar.addfile(ti)
            log.info("'{0}' added".format(f[0]))

    def __etc_crictl_yaml(self,tar):
        data = (self.__tpl_critcl_yaml.render() + "\n").encode()
        fname = "etc/crictl.yaml"
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __flatcar_update_conf(self,tar):
        data = (self.__tpl_flatcar_update_conf.render() + "\n").encode()
        fname = "etc/flatcar/update.conf"
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __etc_i12e_iface_txt(self,tar):
        fname = "etc/i12e/iface.txt"
        iface = self.__cfg.get("flannel",{}).get("interface")
        if iface is None:
            log.info("'{0}' not added: flannel.interface not defined".format(fname))
            return
        iface = iface.strip()
        if len(iface) < 1:
            log.info("'{0}' not added: flannel.interface is empty".format(fname))
            return
        data = (iface + "\n").encode()
        finfo = self.__tarinfo(fname)
        finfo.mode = 0o600
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __k3s_config_yaml(self,tar,mode=None):
        if mode is None:
            fname = "etc/rancher/k3s/config.yaml"
            finfo = self.__tarinfo(fname)
            finfo.mode = 0o600
            tar.addfile(finfo)
            log.info("'{0}' added".format(fname))
            return
        # https://docs.k3s.io/installation/configuration
        data = (self.__tpl_k3s_config_yaml.render(
            i12e_mode = mode,
            k3s_token = self.__cfg["k3s_token"],
            k3s_agent_token = self.__cfg["k3s_agent_token"],
            k3s_url = self.__cfg["k3s_url"],
            tls_san = self.__cfg.get("tls_san"),
        ) + "\n").encode()
        fname = "etc/i12e/k3s/config-{0}.yaml".format(mode)
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        finfo.mode = 0o600
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __k3s_override_conf(self,tar,mode=None):
        if mode is None:
            fname = "etc/systemd/system/k3s.service.d/override.conf"
            finfo = self.__tarinfo(fname)
            finfo.mode = 0o644
            tar.addfile(finfo)
            log.info("'{0}' added".format(fname))
            return
        data = (self.__tpl_k3s_override_conf.render(
            i12e_mode = mode,
        ) + "\n").encode()
        fname = "etc/i12e/k3s/override-{0}.conf".format(mode)
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        finfo.mode = 0o644
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __systemd_i12e_conf(self,tar):
        data = (self.__tpl_systemd_i12e_conf.render() + "\n").encode()
        fname = "etc/systemd/system.conf.d/i12e.conf"
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __opt_bin_e(self,tar):
        data = (self.__tpl_opt_bin_e.render() + "\n").encode()
        fname = "opt/bin/e"
        finfo = self.__tarinfo(fname)
        finfo.mode = 0o755
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __artifact_tune_sh(self,tar):
        data = (self.__tpl_artifact_tune.render() + "\n").encode()
        fname = "opt/libexec/i12e/artifact-tune.sh"
        finfo = self.__tarinfo(fname)
        finfo.mode = 0o755
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __etc_i12e_flags_artifact_pulled(self,tar):
        fname = "etc/i12e/flags/artifact-pulled"
        finfo = self.__tarinfo(fname)
        finfo.mode = 0o600
        tar.addfile(finfo)
        log.info("'{0}' added".format(fname))

    def __nftables_conf(self,tar):
        data = (self.__tpl_nftables_conf.render(
            port_knocking = self.__cfg["port_knocking"],
        ) + "\n").encode()
        fname = "etc/nftables.conf"
        finfo = self.__tarinfo(fname)
        finfo.mode = 0o600
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __nftables_service(self,tar):
        data = (self.__tpl_nftables_service.render() + "\n").encode()
        fname = "etc/systemd/system/nftables.service"
        finfo = self.__tarinfo(fname)
        finfo.size = len(data)
        tar.addfile(finfo,BytesIO(data))
        log.info("'{0}' added".format(fname))

    def __devel_output(self,buf):
        log.info("sha256: {0}".format(sha256(buf).hexdigest()))
        cmd = ['tar','tvz']
        log.info("$ {0}".format(" ".join(cmd)))
        p = Popen(args=cmd,stdin=PIPE)
        p.communicate(buf)
        if p.returncode != 0:
            raise Exception("'{0}' command failed".format(" ".join(cmd)))

    def __rclone_push(self,buf):
        rclone_config = "/root/rclone.conf"
        with open(rclone_config,"w") as fp:
            fp.write(Config().rclone_config())
        environ["RCLONE_CONFIG"] = rclone_config   # after encrypted one has been read
        # --------
        cmd = ['rclone','rcat','rem:artifact.tar.gz']
        log.info("$ {0}".format(" ".join(cmd)))
        p = Popen(args=cmd,stdin=PIPE)
        p.communicate(buf)
        if p.returncode != 0:
            raise Exception("'{0}' command failed".format(" ".join(cmd)))
        # --------
        cmd = ['rclone','cat','rem:artifact.tar.gz']
        log.info("$ {0}".format(" ".join(cmd)))
        p = Popen(args=cmd,stdout=PIPE)
        (odata,_) = p.communicate()
        if p.returncode != 0:
            raise Exception("'{0}' command failed".format(" ".join(cmd)))
        # --------
        sha256_1 = sha256(buf).hexdigest()
        sha256_2 = sha256(odata).hexdigest()
        log.info("sha256(bef): {0}".format(sha256_1))
        log.info("sha256(aft): {0}".format(sha256_2))
        if sha256_1 != sha256_2:
            raise Exception("sha256 checksum mismatch")
        # --------
        cmd = ['tar','tvz']
        log.info("$ {0}".format(" ".join(cmd)))
        p = Popen(args=cmd,stdin=PIPE)
        p.communicate(odata)
        if p.returncode != 0:
            raise Exception("'{0}' command failed".format(" ".join(cmd)))

    def run(self):
        log.info("==== genesis artifact begin ====")
        buf = BytesIO()
        with tar_open(fileobj=buf, mode="w:gz") as tar:
            self.__folders(tar)
            self.__etc_crictl_yaml(tar)
            self.__flatcar_update_conf(tar)
            self.__etc_i12e_iface_txt(tar)
            for mode in self.__modes:
                self.__k3s_config_yaml(tar,mode)
            for mode in self.__modes:
                self.__k3s_override_conf(tar,mode)
            self.__nftables_conf(tar)
            self.__systemd_i12e_conf(tar)
            self.__k3s_override_conf(tar)
            self.__nftables_service(tar)
            self.__k3s_config_yaml(tar)
            self.__opt_bin_e(tar)
            self.__artifact_tune_sh(tar)
            self.__etc_i12e_flags_artifact_pulled(tar)
        if self.__devel:
            self.__devel_output(buf.getvalue())
            return
        self.__rclone_push(buf.getvalue())
        log.info("---- genesis artifact end ----")

File: internal/artifact/artifact.go
package artifact

import (
	"archive/tar"
	"bytes"
	"compress/gzip"
	"crypto/sha256"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/sfmunoz/i12e/internal/cmdutil"
	"github.com/sfmunoz/i12e/internal/config"
	"github.com/sfmunoz/logit"
)

var log = logit.Logit().WithLevel(logit.LevelInfo)

type Artifact struct {
	cfg    *config.ArtifactConfig
	tnow   time.Time
	tarOut *tar.Writer
	gzOut  *gzip.Writer
	obuf   *bytes.Buffer
}

func newArtifact(cfg *config.ArtifactConfig) (*Artifact, error) {
	if cfg == nil {
		return nil, fmt.Errorf("newArtifact(): undefined config")
	}
	var obuf bytes.Buffer
	gzOut := gzip.NewWriter(&obuf)
	tarOut := tar.NewWriter(gzOut)
	return &Artifact{
		cfg:    cfg,
		tnow:   time.Now().UTC().Truncate(time.Second), // Truncate(): to prevent 'timestamp in the future' warnings
		tarOut: tarOut,
		gzOut:  gzOut,
		obuf:   &obuf,
	}, nil
}

func (a *Artifact) Close() error {
	if err := a.tarOut.Close(); err != nil {
		return err
	}
	if err := a.gzOut.Close(); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) folders() error {
	flist := []struct {
		Name string
		Mode int64
	}{
		{Name: "etc/i12e", Mode: 0700},
		{Name: "etc/i12e/flags", Mode: 0700},
		{Name: "etc/i12e/k3s", Mode: 0700},
		{Name: "etc/i12e/flux", Mode: 0700},
		{Name: "etc/systemd/system.conf.d", Mode: 0755},
		{Name: "etc/wireguard", Mode: 0700}, // it's ok and harmless: it already exists like this
		{Name: "opt/libexec", Mode: 0755},
		{Name: "opt/libexec/i12e", Mode: 0755},
		{Name: "opt/libexec/i12e/plugins", Mode: 0755},
	}
	for _, folder := range flist {
		hdr := &tar.Header{
			Typeflag: tar.TypeDir,
			Name:     folder.Name,
			ModTime:  a.tnow,
			Mode:     folder.Mode,
			Uid:      0,
			Gid:      0,
			Uname:    "root",
			Gname:    "root",
		}
		if err := a.tarOut.WriteHeader(hdr); err != nil {
			return err
		}
	}
	return nil
}

func (a *Artifact) addStatic(staticFname, targetFname string, mode int64) error {
	body, err := FS.ReadFile(staticFname)
	if err != nil {
		return err
	}
	hdr := &tar.Header{
		Typeflag: tar.TypeReg,
		Name:     targetFname,
		ModTime:  a.tnow,
		Size:     int64(len(body)),
		Mode:     mode,
		Uid:      0,
		Gid:      0,
		Uname:    "root",
		Gname:    "root",
	}
	if err := a.tarOut.WriteHeader(hdr); err != nil {
		return err
	}
	if _, err := a.tarOut.Write(body); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) addEmpty(targetFname string, mode int64) error {
	hdr := &tar.Header{
		Typeflag: tar.TypeReg,
		Name:     targetFname,
		ModTime:  a.tnow,
		Size:     0,
		Mode:     mode,
		Uid:      0,
		Gid:      0,
		Uname:    "root",
		Gname:    "root",
	}
	if err := a.tarOut.WriteHeader(hdr); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) addTemplate(templateFname, targetFname string, mode int64, data any) error {
	tpl, err := tplNew(templateFname, true)
	if err != nil {
		return err
	}
	var body bytes.Buffer
	err = tpl.Execute(&body, data)
	if err != nil {
		return err
	}
	hdr := &tar.Header{
		Typeflag: tar.TypeReg,
		Name:     targetFname,
		ModTime:  a.tnow,
		Size:     int64(body.Len()),
		Mode:     mode,
		Uid:      0,
		Gid:      0,
		Uname:    "root",
		Gname:    "root",
	}
	if err := a.tarOut.WriteHeader(hdr); err != nil {
		return err
	}
	if _, err := a.tarOut.Write(body.Bytes()); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcCrictlYaml() error {
	if err := a.addStatic("static/crictl.yaml", "etc/crictl.yaml", 0644); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcFlatcarUpdateConf() error {
	if err := a.addStatic("static/flatcar-update.conf", "etc/flatcar/update.conf", 0644); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcI12eIfaceTxt() error {
	targetName := "etc/i12e/iface.txt"
	if len(a.cfg.Mesh.EndpointInterface) < 1 {
		log.Info("skipping: undefined 'mesh.endpoint_interface'", "targetName", targetName)
		return nil
	}
	body := []byte(a.cfg.Mesh.EndpointInterface + "\n")
	hdr := &tar.Header{
		Typeflag: tar.TypeReg,
		Name:     targetName,
		ModTime:  a.tnow,
		Size:     int64(len(body)),
		Mode:     0600,
		Uid:      0,
		Gid:      0,
		Uname:    "root",
		Gname:    "root",
	}
	if err := a.tarOut.WriteHeader(hdr); err != nil {
		return err
	}
	if _, err := a.tarOut.Write(body); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcI12eK3sConfigYaml() error {
	data := struct {
		I12eMode      string
		K3sToken      string
		K3sAgentToken string
		K3sUrl        string
		TlsSan        string
	}{
		I12eMode:      "",
		K3sToken:      a.cfg.Artifact.K3sToken,
		K3sAgentToken: a.cfg.Artifact.K3sAgentToken,
		K3sUrl:        fmt.Sprintf("https://%s:6443", a.cfg.K3s.TlsSan),
		TlsSan:        a.cfg.K3s.TlsSan,
	}
	for _, m := range config.ValidModes() {
		data.I12eMode = m
		targetName := fmt.Sprintf("etc/i12e/k3s/config-%s.yaml", m)
		if err := a.addTemplate("k3s-config.yaml", targetName, 0600, &data); err != nil {
			return err
		}
	}
	return nil
}

func (a *Artifact) etcI12eK3sOverrideConf() error {
	for _, m := range config.ValidModes() {
		data := struct{ I12eMode string }{I12eMode: m}
		targetName := fmt.Sprintf("etc/i12e/k3s/override-%s.conf", m)
		if err := a.addTemplate("k3s-override.conf", targetName, 0644, &data); err != nil {
			return err
		}
	}
	return nil
}

func (a *Artifact) etcI12eFluxCfg() error {
	data := struct {
		Cluster     string
		GitHubToken string
		AgeKey      string
	}{
		Cluster:     a.cfg.Env.String(),
		GitHubToken: a.cfg.Flux.GitHubToken,
		AgeKey:      a.cfg.Flux.AgeKey,
	}
	return a.addTemplate("flux.cfg", "etc/i12e/flux/flux.cfg", 0600, &data)
}

func (a *Artifact) etcNftablesConf() error {
	data := struct {
		PortKnocking []int
		EndpointPort int
	}{
		PortKnocking: a.cfg.Artifact.PortKnocking,
		EndpointPort: a.cfg.Mesh.EndpointPort,
	}
	if err := a.addTemplate("nftables.conf", "etc/nftables.conf", 0600, &data); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcSystemdSystemConfDI12eConf() error {
	if err := a.addStatic("static/systemd-i12e.conf", "etc/systemd/system.conf.d/i12e.conf", 0644); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcSystemdSystemNftablesService() error {
	if err := a.addStatic("static/nftables.service", "etc/systemd/system/nftables.service", 0644); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) etcWireguard() error {
	if len(a.cfg.WgConf) < 1 {
		log.Info("skipping: undefined 'wg_conf' array")
		return nil
	}
	for i, v := range a.cfg.WgConf {
		targetName := fmt.Sprintf("etc/wireguard/wg%d.conf", i)
		body := []byte(v)
		hdr := &tar.Header{
			Typeflag: tar.TypeReg,
			Name:     targetName,
			ModTime:  a.tnow,
			Size:     int64(len(body)),
			Mode:     0600,
			Uid:      0,
			Gid:      0,
			Uname:    "root",
			Gname:    "root",
		}
		if err := a.tarOut.WriteHeader(hdr); err != nil {
			return err
		}
		if _, err := a.tarOut.Write(body); err != nil {
			return err
		}
	}
	return nil
}

func (a *Artifact) optBinE() error {
	if err := a.addStatic("static/opt-bin-e", "opt/bin/e", 0755); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) optLibexecI12eArtifactTuneSh() error {
	if err := a.addStatic("static/artifact-tune.sh", "opt/libexec/i12e/artifact-tune.sh", 0755); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) optLibexecI12eK3sInstallSh() error {
	if err := a.addStatic("static/k3s-install.sh", "opt/libexec/i12e/k3s-install.sh", 0755); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) optLibexecI12ePlugins() error {
	const pluginsDir = "plugins"
	entries, err := os.ReadDir(pluginsDir)
	if err != nil {
		if os.IsNotExist(err) {
			log.Info("skipping: 'plugins/' folder not found")
			return nil
		}
		return err
	}
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		srcPath := filepath.Join(pluginsDir, entry.Name())
		body, err := os.ReadFile(srcPath)
		if err != nil {
			return err
		}
		targetName := "opt/libexec/i12e/plugins/" + entry.Name()
		hdr := &tar.Header{
			Typeflag: tar.TypeReg,
			Name:     targetName,
			ModTime:  a.tnow,
			Size:     int64(len(body)),
			Mode:     0644,
			Uid:      0,
			Gid:      0,
			Uname:    "root",
			Gname:    "root",
		}
		if err := a.tarOut.WriteHeader(hdr); err != nil {
			return err
		}
		if _, err := a.tarOut.Write(body); err != nil {
			return err
		}
	}
	return nil
}

func (a *Artifact) etcI12eFlagsArtifactPulled() error {
	if err := a.addEmpty("etc/i12e/flags/artifact-pulled", 0600); err != nil {
		return err
	}
	return nil
}

func (a *Artifact) rclonePush() error {
	remFile := fmt.Sprintf("%s:artifact.tar.gz", a.cfg.Rclone.Remote)
	log.Info("rclonePush()", "remFile", remFile)
	// --------
	sha256_1 := fmt.Sprintf("%x", sha256.Sum256(a.obuf.Bytes())) // keep it before buffer read
	// --------
	cmd1 := exec.Command("rclone", "rcat", remFile)
	cmd1.Stdin = a.obuf
	bo1, be1, err1 := cmdutil.RunSimple(cmd1)
	if err1 != nil {
		return fmt.Errorf("'rclone rcat' failed': %s (stdout=%s, stderr=%s)", err1, bo1.String(), be1.String())
	}
	// --------
	cmd2 := exec.Command("rclone", "cat", remFile)
	bo2, be2, err2 := cmdutil.RunSimple(cmd2)
	if err2 != nil {
		return fmt.Errorf("'rclone cat' failed': %s (stdout=%s, stderr=%s)", err2, bo2.String(), be2.String())
	}
	// --------
	sha256_2 := fmt.Sprintf("%x", sha256.Sum256(bo2.Bytes()))
	log.Info("sha256(bef)", "sha256", sha256_1)
	log.Info("sha256(aft)", "sha256", sha256_2)
	if sha256_1 != sha256_2 {
		return fmt.Errorf("sha256 checksum mismatch")
	}
	// --------
	cmd3 := exec.Command("tar", "tvz")
	cmd3.Stdin = bo2
	bo3, be3, err3 := cmdutil.RunSimple(cmd3)
	if err3 != nil {
		return fmt.Errorf("'tar tvz' failed': %s (stdout=%s, stderr=%s)", err3, bo3.String(), be3.String())
	}
	for _, line := range strings.Split(bo3.String(), "\n") {
		log.Info(">", "tgz", line)
	}
	return nil
}

func (a *Artifact) run() error {
	funcList := []func() error{
		a.folders,
		a.etcCrictlYaml,
		a.etcFlatcarUpdateConf,
		a.etcI12eIfaceTxt,
		a.etcI12eK3sConfigYaml,
		a.etcI12eK3sOverrideConf,
		a.etcI12eFluxCfg,
		a.etcNftablesConf,
		a.etcSystemdSystemConfDI12eConf,
		a.etcSystemdSystemNftablesService,
		a.etcWireguard,
		a.optBinE,
		a.optLibexecI12eArtifactTuneSh,
		a.optLibexecI12eK3sInstallSh,
		a.optLibexecI12ePlugins,
		a.etcI12eFlagsArtifactPulled,
	}
	for _, f := range funcList {
		if err := f(); err != nil {
			a.Close()
			return err
		}
	}
	if err := a.Close(); err != nil {
		return err
	}
	if err := a.rclonePush(); err != nil {
		return err
	}
	return nil
}

func Run(cfg *config.ArtifactConfig) error {
	a, err := newArtifact(cfg)
	if err != nil {
		return err
	}
	return a.run()
}
